handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java

```diff
@@ -596,11 +602,13 @@ public final SSLEngineResult wrap(
 
                 status = handshake();
                 if (status == NEED_UNWRAP) {
-                    return NEED_UNWRAP_OK;
+                    // Signal if the outbound is done or not.
+                    return isOutboundDone() ? NEED_UNWRAP_CLOSED : NEED_UNWRAP_OK;
                 }
 
-                if (engineClosed) {
-                    return NEED_UNWRAP_CLOSED;
+                if (outboundClosed) {
```

Comment:
consider adding a comment here that we explicitly want to use `outboundClosed` instead of calling `outboundClosed()` because we want to drain any bytes if they are available.